wcc-contentful/lib/wcc/contentful/store/query.rb

```diff
@@ -0,0 +1,255 @@
+        apply_operator(op, field, expected, context)
+      end
+    end
+
+    # Called with a filter object by {Base#find_by} in order to apply the filter.
+    # The filter in this case is a hash where the keys are paths and the values
+    # are expectations.
+    # @see #apply_operator
+    def apply(filter, context = nil)
+      self.class.flatten_filter_hash(filter).reduce(self) do |query, cond|
+        query.apply_operator(cond[:op], cond[:path], cond[:expected], context)
+      end
+    end
+
+    protected
+
+    def _append_condition(condition)
+      self.class.new(
+        store,
+        content_type: content_type,
+        conditions: conditions + [condition],
+        options: @options,
+        **@extra
+      )
+    end
+
+    # naive implementation recursively descends the graph to turns links into
+    # the actual entry data.  This calls {Base#find} for each link and so it is
+    # very inefficient.
+    #
+    # @abstract Override this to provide a more efficient implementation for
+    #   a given store.
+    def resolve_includes(entry, depth)
+      return entry unless entry && depth && depth > 0
+
+      WCC::Contentful::LinkVisitor.new(entry, :Link, :Asset, depth: depth).map! do |val|
+        resolve_link(val)
+      end
+    end
+
+    def resolve_link(val)
+      return val unless val.is_a?(Hash) && val.dig('sys', 'type') == 'Link'
+      return val unless included = @store.find(val.dig('sys', 'id'))
+
+      included
+    end
+
+    class << self
+      def op?(key)
+        OPERATORS.include?(key.to_sym)
+      end
+
+      # Turns a hash into a flat array of individual conditions, where each
+      # element can be passed as params to apply_operator
+      def flatten_filter_hash(hash, path = [])
+        hash.flat_map do |(k, v)|
+          k = k.to_s
+          if k.include?('.')
+            k, *rest = k.split('.')
+            v = { rest.join('.') => v }
+          end
+
+          if v.is_a? Hash
+            flatten_filter_hash(v, path + [k])
+          elsif op?(k)
+            { path: path, op: k.to_sym, expected: v }
+          else
+            { path: path + [k], op: :eq, expected: v }
+          end
+        end
+      end
+
+      def known_locales
+        @known_locales = WCC::Contentful.locales.keys
+      end
+      RESERVED_NAMES = %w[fields sys].freeze
+
+      # Takes a path array in non-normal form and inserts 'sys', 'fields',
+      # and the current locale as appropriate to normalize it.
+      # rubocop:disable Metrics/BlockNesting
+      def normalize_condition_path(path, context = nil)
+        context_locale = context[:locale] if context.present?
+        context_locale ||= 'en-US'
+
+        rev_path = path.reverse
+        new_path = []
+
+        current_tuple = []
+        current_locale_was_inferred = false
+        until rev_path.empty? && current_tuple.empty?
+          raise ArgumentError, "Query too complex: #{path.join('.')}" if new_path.length > 7
```

Comment:
is this enforcing no more than one layer of nested querying?

Reply:
Yep